File: app/src/main/java/xdman/ui/components/SegmentPanel.java
/*
 * Copyright (c)  Subhra Das Gupta
 *
 * This file is part of Xtreme Download Manager.
 *
 * Xtreme Download Manager is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Xtreme Download Manager is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with Xtream Download Manager; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 * 
 */

package xdman.ui.components;

import java.awt.Color;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.util.ArrayList;

import javax.swing.JComponent;

import xdman.downloaders.SegmentDetails;
import xdman.downloaders.SegmentInfo;
import xdman.ui.res.ColorResource;

public class SegmentPanel extends JComponent {

	private static final long serialVersionUID = -6537879808121349569L;
	private SegmentDetails segDet;
	private long length;

	public void setValues(SegmentDetails segDet, long length) {
		this.segDet = segDet;
		this.length = length;
		repaint();
	}

	public void paintComponent(Graphics g) {
		if (g == null)
			return;

		Graphics2D g2 = (Graphics2D) g;
		g2.setPaint(Color.GRAY);
		g2.fillRect(0, 0, getWidth(), getHeight());
		if (segDet == null || segDet.getChunkCount() < 1 || length < 0) {
			return;
		}

		g2.setPaint(ColorResource.getSelectionColor());

		float r = (float) getWidth() / length;

		ArrayList<SegmentInfo> list = segDet.getChunkUpdates();
		for (int i = 0; i < segDet.getChunkCount(); i++) {
			SegmentInfo info = list.get(i);
			int _start = (int) (info.getStart() * r);
			int _length = (int) (info.getLength() * r);
			int _dwnld = (int) (info.getDownloaded() * r);
			if (_dwnld > _length)
				_dwnld = _length;
			g2.fillRect(_start, 0, _dwnld + 1, getHeight());

		}
	}

}

File: xdm-app/src/main/java/xdm/app/ui/components/SegmentPanel.kt
/*
 * Copyright (c)  Subhra Das Gupta
 *
 * This file is part of Xtream Download Manager.
 *
 * Xtream Download Manager is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Xtream Download Manager is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with Xtream Download Manager; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */
package xdm.app.ui.components

import xdm.core.downloaders.web.SegmentProgress
import java.awt.Color
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.geom.RoundRectangle2D
import javax.swing.JComponent
import javax.swing.UIManager

class SegmentPanel : JComponent() {
    private var segDet: Collection<SegmentProgress> = listOf()
    private var totalSize: Long = 0

    fun setValues(segDet: Collection<SegmentProgress>) {
        this.segDet = segDet
        this.totalSize = segDet.sumOf { it.length }
        repaint()
    }

    public override fun paintComponent(g: Graphics) {
        val g2 = g as Graphics2D
        g2.paint = UIManager.getColor("ProgressBar.background")
        g2.fillRect(0, 0, getWidth(), getHeight())
        val originalClip = g2.clip
        val clipShape = RoundRectangle2D.Double(0.0, 0.0, getWidth().toDouble(), getHeight().toDouble(), 5.0, 5.0)
        g2.clip = clipShape
        try {
//            g2.paint = Color.GRAY
            g2.fillRect(0, 0, getWidth(), getHeight())
            if (segDet.isEmpty() || totalSize < 0) {
                return
            }

            g2.paint = UIManager.getColor("ProgressBar.foreground")

            val r = getWidth().toFloat() / totalSize
            for (info in segDet) {
                val start = (info.start * r).toInt()
                val length = (info.length * r).toInt()
                var downloaded = (info.downloaded * r).toInt()
                if (downloaded > length) downloaded = length
                g2.fillRect(start, 0, downloaded + 1, getHeight())
            }
        } finally {
            g2.clip = originalClip
        }
    }
}
